src/lib/ERC4337/validator/Validator.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import {IValidator} from "src/lib/ERC4337/validator/interface/IValidator.sol";
+import {UserOperation} from "src/lib/ERC4337/utils/UserOperation.sol";
+import {Ownable} from "src/access/ownable/Ownable.sol";
+import {ECDSA} from "openzeppelin-contracts/utils/cryptography/ECDSA.sol";
+
+abstract contract Validator is IValidator {
+
+    /*=============
+        STORAGE
+    =============*/
+
+    /// @dev Error code for invalid EIP-4337 `validateUserOp()`signature
+    /// @dev Error return value is abbreviated to 1 since it need not include time range
+    uint8 internal constant SIG_VALIDATION_FAILED = 1;
+    /// @dev Error code for invalid EIP-1271 signature in `isValidSignature()`
+    /// @dev Nonzero to define invalid sig error, as opposed to wrong validator address error, ie: `bytes4(0)`
+    bytes4 internal constant INVALID_SIGNER = hex'ffffffff';
+
+    /// @dev Since the EntryPoint contract uses chainid and its own address to generate request ids, 
+    /// its address on this chain must be available to all ERC4337-compliant validators.
+    address public immutable entryPoint;
+
+    constructor(address _entryPointAddress) {
+        entryPoint = _entryPointAddress; 
+    }
+
+    /*===============
+        VALIDATOR
+    ===============*/
+
+    /// @dev Convenience function to generate an EntryPoint request id for a given UserOperation.
+    /// Use this output to generate an un-typed digest for signing to comply with `eth_sign` + EIP-191 
+    /// @param userOp The 4337 UserOperation to hash. The struct's signature member is discarded.
+    /// @notice Can also be done offchain or called directly on the EntryPoint contract as it is identical
+    function getUserOpHash(UserOperation calldata userOp) public view returns (bytes32) {
+        return keccak256(abi.encode(_innerOpHash(userOp), address(entryPoint), block.chainid));
+    }
+
+    /*===============
+        INTERNALS
+    ===============*/
+
+    /// @dev Function to verify signatures forwarded to this contract. Must be overridden by validator
+    /// child contracts in a way that suits the use case of their design
+    function _verifySignature(
+        address signer, 
+        bytes32 userOpHash, 
+        bytes memory nestedSignature
+    ) internal view virtual returns (bool);

```

Comment:
While our current validator use cases use this schema, it might not apply to all Validators. For example, what if Station runs a Validator that checks if the signature comes from Station's central BotAccount. The `signer` argument is needed because it'll be read from state and managed by our multisig. I'd opt to remove this for now

Reply:
Resolved in 4c988f9

In reworking this I remembered that the only reason I included the signer address was to avoid performing `ECDSA.recover()` twice, since `SignatureChecker.isValidSignatureNow()` invokes it under the hood and EIP1271's `isValidSignature()` doesn't provide a signer address as part of the call context.

With the signer address param removed, I moved signature recovery up into the external function calls (`validateUserOp()` and `isValidSignature()`) and used the signature recovery method appropriate for the call context. Those functions are expected to have signature validation in their scopes anyway.

This also came with the benefit of `_verifySignature()` now only needing to verify the `CALL_PERMIT` permission for the recovered signer address, so I gave it a more accurate name for its new, tighter scope: `_verifySigner()`. I like this a lot because now contracts that inherit `Validator` only need to think about implementing access control authentication in `_verifySigner()`